From 429b200aa0e21b33a720790d18e059491eb60089 Mon Sep 17 00:00:00 2001
From: Daniel Drake <dsd@laptop.org>
Date: Wed, 02 Mar 2011 15:52:53 +0000
Subject: Disable keybindings message should ungrab keys

The disable-keybindings action is currently only 'soft' in that it
just causes Metacity to ignore keypresses; the keys themselves are left
bound.

Sugar would like to use this message to actually disable the bindings
*and* release the grabs so that it can implement its own Alt-Tab handler.

This patch implements this. Based on earlier work by Bernie Innocenti.
---
diff -ur metacity-2.30.3-ylmf1/src/core/display.c metacity-2.30.3-ylmf1.new//src/core/display.c
--- metacity-2.30.3-ylmf1/src/core/display.c	2011-09-30 10:00:47.965830780 +0800
+++ metacity-2.30.3-ylmf1.new//src/core/display.c	2011-09-30 10:01:35.962496812 +0800
@@ -2461,7 +2461,7 @@
                 {
                   meta_verbose ("Received set keybindings request = %d\n",
                                 (int) event->xclient.data.l[0]);
-                  meta_set_keybindings_disabled (!event->xclient.data.l[0]);
+		  meta_set_keybindings_disabled (display, !event->xclient.data.l[0]);		
                 }
               else if (event->xclient.message_type ==
                        display->atom__METACITY_TOGGLE_VERBOSE)
diff -ur metacity-2.30.3-ylmf1/src/core/keybindings.c metacity-2.30.3-ylmf1.new//src/core/keybindings.c
--- metacity-2.30.3-ylmf1/src/core/keybindings.c	2011-09-30 10:00:47.669164118 +0800
+++ metacity-2.30.3-ylmf1.new//src/core/keybindings.c	2011-09-30 10:04:12.299161406 +0800
@@ -768,7 +768,10 @@
 
   if (screen->keys_grabbed)
     return;
-
+  
+  if (all_bindings_disabled)
+     return;
+  
   grab_keys (screen->display->key_bindings,
              screen->display->n_key_bindings,
              screen->display, screen->xroot,
@@ -792,7 +795,10 @@
 {
   if (window->all_keys_grabbed)
     return;
-
+  
+  if (all_bindings_disabled)
+      return;
+  
   if (window->type == META_WINDOW_DOCK)
     {
       if (window->keys_grabbed)
@@ -3323,9 +3329,11 @@
 }
 
 void
-meta_set_keybindings_disabled (gboolean setting)
+meta_set_keybindings_disabled (MetaDisplay *display,
+                               gboolean     setting)
 {
   all_bindings_disabled = setting;
+  regrab_key_bindings (display);
   meta_topic (META_DEBUG_KEYBINDINGS,
               "Keybindings %s\n", all_bindings_disabled ? "disabled" : "enabled");
 }
diff -ur metacity-2.30.3-ylmf1/src/core/keybindings.h metacity-2.30.3-ylmf1.new//src/core/keybindings.h
--- metacity-2.30.3-ylmf1/src/core/keybindings.h	2010-09-05 00:09:53.000000000 +0800
+++ metacity-2.30.3-ylmf1.new//src/core/keybindings.h	2011-09-30 10:04:38.765827723 +0800
@@ -49,7 +49,9 @@
 void     meta_display_process_key_event     (MetaDisplay *display,
                                              MetaWindow  *window,
                                              XEvent      *event);
-void     meta_set_keybindings_disabled      (gboolean     setting);
+void     meta_set_keybindings_disabled      (MetaDisplay *display,
+                                             gboolean     setting);
+
 void     meta_display_process_mapping_event (MetaDisplay *display,
                                              XEvent      *event);
 
